### DISCUSSION TITLE
How to wait or sleep in a component?

### DISCUSSION THREAD

**Discussion by sheershak-agarwal** [nasa/fprime#1014] [Q&A]
Probably a dumb question but I am trying to sleep in my gps component after resetting it. My Gps Component is also connected to the Health Ping, so when I sleep it stops replying to those pings.

Is there a better way to delay to wait in fprime without missing the pings?

**Comment by timcanham** [comment]
It's actually a good question. If by "sleep" you mean wait a specified amount of time for the reset to complete so you can talk to it, I usually decrement a counter in the scheduler call until it reaches the time I want to wait, then start talking again. That way the component execution won't block waiting for the time to pass.

**Reply by sheershak-agarwal** [reply]
I was using this 
```C++
clock_t endtime;
endtime = clock();
endtime+= data->seconds * CLOCKS_PER_SEC;
while (clock()<endtime) {}
```

Bit I can still see the error:
Assert file "/home/sheershak/Desktop/F-Prime-Debug/App/build-fprime-automatic-raspberrypi/App/Gps/GpsComponentAc.cpp": 
Assert file "/home/sheershak/Desktop/F-Prime-Debug/App/build-fprime-automatic-raspberrypi/App/Gps/GpsComponentAc.cpp": 
EVENT: (121) (2:1630765741,969675) WARNING_HI: (health) HLTH_PING_WARN : Ping entry GPS late warning
Assert file "/home/sheershak/Desktop/F-Prime-Debug/App/build-fprime-automatic-raspberrypi/App/Gps/GpsComponentAc.cpp": 
Assert file "/home/sheershak/Desktop/F-Prime-Debug/App/build-fprime-automatic-raspberrypi/App/Gps/GpsComponentAc.cpp":

**Reply by sheershak-agarwal** [reply]
Even if I remove the Ping for my component. I can still see the Assert File line but it doesn't say  Ping entry GPS late warning or the FATAL error. which makes me think the assert file error is coming from something else.

**Comment by timcanham** [comment]
This is how you do what I suggested.

Create a private variable in your component:

```c++
U32 m_resetCntDown;
```

Assume for the sake of discussion that you call your rate group port 1/sec, and that it takes the GPS 5 seconds to fully reset.

When you command the reset, you do this:

```c++
m_resetCntDown = 5
```

Each time the rate group gets called, you do:

```
m_resetCntDown--;
```

In the same call, you check:

```c++
if (m_resetCntDown == 0) {
    // start operating GPS again
}
```

That way, your component never blocks.
